lxd/acme/acme.go

```diff
@@ -0,0 +1,179 @@
+			return nil, fmt.Errorf("Failed to parse certificate: %w", err)
+		}
+		// If the domain matches the certificate's DNS names, and the certificate is valid for more
+		// than 30 days, return it.
+		if shared.StringInSlice(domain, cert.DNSNames) && !time.Now().After(cert.NotAfter.Add(-time.Hour*24*30)) {
+			return &certificate.Resource{
+				Certificate: clusterCert,
+				PrivateKey:  key,
+			}, nil
+		}
+	}
+
+	// Load the certificate.
+	certInfo, err := util.LoadCert(s.OS.VarDir)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load certificate and key file: %w", err)
+	}
+
+	cert, err := x509.ParseCertificate(certInfo.KeyPair().Certificate[0])
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse certificate: %w", err)
+	}
+
+	// Check the certificate's expiration date, and renew it if it's only valid for less than 30 days.
+	// This is only done if our certificate's DNS names contains the request domain.
+	if shared.StringInSlice(domain, cert.DNSNames) && !time.Now().After(cert.NotAfter.Add(-time.Hour*24*30)) {
+		logger.Debug("Skipping certificate renewal as it is still valid for more than 30 days")
+		return nil, nil
+	}
+
+	// Generate new private key for user. This key needs to be different from the server's private key.
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("Failed generating private key for user account: %w", err)
+	}
+
+	user := user{
+		Email: email,
+		Key:   privateKey,
+	}
+
+	config := lego.NewConfig(&user)
+
+	if caURL != "" {
+		config.CADirURL = caURL
+	} else {
+		// Default URL for Let's Encrypt
+		config.CADirURL = "https://acme-v02.api.letsencrypt.org/directory"
+	}
+
+	config.Certificate.KeyType = certcrypto.RSA2048
+
+	client, err := lego.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create new client: %w", err)
+	}
+
+	err = client.Challenge.SetHTTP01Provider(provider)
+	if err != nil {
+		return nil, fmt.Errorf("Failed setting HTTP-01 provider: %w", err)
+	}
+
+	var reg *registration.Resource
+
+	// Registration might fail randomly (as seen in manual tests), so retry in that case.
+	for i := 0; i < retries; i++ {
+		reg, err = client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
+		if err == nil {
+			break
+		}
+
+		// In case we were rate limited, don't try again.
+		details, ok := err.(*acme.ProblemDetails)
+		if ok && details.Type == "urn:ietf:params:acme:error:rateLimited" {
+			break
+		}
+
+		logger.Warn("Failed to register user, retrying in 10 seconds", logger.Ctx{"err": err})
+		time.Sleep(10 * time.Second)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("Failed to register user: %w", err)
+	}
+
+	user.Registration = reg
+
+	request := certificate.ObtainRequest{
+		Domains:    []string{domain},
+		Bundle:     true,
+		PrivateKey: certInfo.KeyPair().PrivateKey,
+	}
+
+	var certificates *certificate.Resource
+
+	logger.Info("Issuing certificate")
```

Comment:
Can you add a new context to this logger with the domain name to all log messages in this function?